packages/dds/sequence/src/sequence.ts

```diff
@@ -233,14 +233,18 @@ export abstract class SharedSegmentSequence<T extends ISegment>
 	 * @param end - The exclusive end of the range to remove
 	 */
 	public removeRange(start: number, end: number): IMergeTreeRemoveMsg {
-		const removeOp = this.client.removeRangeLocal(start, end);
-		this.submitSequenceMessage(removeOp);
-		return removeOp;
+		return ensureNoReentrancy(() => {
+			const removeOp = this.client.removeRangeLocal(start, end);
+			this.submitSequenceMessage(removeOp);
+			return removeOp;
+		});
 	}
 
 	public groupOperation(groupOp: IMergeTreeGroupMsg) {
```

Comment:
this seems redundant with the reentrancy protection that @andre4i built at the runtime level. Outside of group ops is there actually a danger with reentrancy? if it is only group ops, i think we should just deprecate them, and stop using them internally, as they are from a time without any batching and don't provide any additional guarantees beyond batching

Reply:
actually, now that i think about it, it does seem like there is more than group ops, due to how we hook up eventing, and submission all operations done in an event will end up submitted before the operations that create the event. I also think we could fix this by moving the submit before the emit in the client's delta handler, and remove the direct submits from each edit method

Reply:
1. The runtime level reentrancy isn't actually enabled as it's extensively used (a bit scary, but not necessarily the worst due to 2)
2. This is a different beast of reentrancy than what that checks: the runtime reentrancy verifies that processing a sequenced op doesn't incur submission of further ops. The problem cases here are where submitting a local op incurs submission of further local ops.

There are some things we could do to refactor sequence/client's interaction to make this type of reentrancy less scary, but it's a more invasive change and probably requires public API changes.

Reply:
Notably there aren't any group ops involved at all in the test case I wrote, which is relatively faithful to what happened on the corrupted doc I investigated, just a bit simplified

Reply:
if you replace the `assert.throws` in that test case with a synchronize + assert that the shared strings match (i.e. the claim that we handle this kind of reentrancy), you hit 0x046 on the client submitting the reentrant ops, and 'MergeTree insert failed' on the second string (not that the test gets this far--have to do some muddling in the debugger to see that)

Reply:
I worry this is too disruptive of a change; specifically it will break out partners apps. i think we can fix it under the covers.